packages/app/src/background/services/connection/ConnectionService.ts

```diff
@@ -0,0 +1,170 @@
+import SimpleStorage from "@src/background/services/storage";
+import ZkIdentityService from "@src/background/services/zkIdentity";
+
+import type { IIdenityConnection, IZkMetadata, IConnectArgs, IIdentityMetadata } from "@cryptkeeperzk/types";
+
+const CONNECTIONS_STORAGE_KEY = "@@CONNECTIONS@@";
+
+export default class ConnectionService extends BaseService implements IBackupable {
+  private static INSTANCE?: ConnectionService;
+
+  private connections: Map<string, IIdenityConnection>;
+
+  private readonly connectionsStorage: SimpleStorage;
+
+  private readonly cryptoService: CryptoService;
+
+  private readonly browserController: BrowserUtils;
+
+  private readonly zkIdenityService: ZkIdentityService;
+
+  private constructor() {
+    super();
+    this.connections = new Map();
+    this.connectionsStorage = new SimpleStorage(CONNECTIONS_STORAGE_KEY);
+    this.cryptoService = CryptoService.getInstance();
+    this.browserController = BrowserUtils.getInstance();
+    this.zkIdenityService = ZkIdentityService.getInstance();
+  }
+
+  static getInstance = (): ConnectionService => {
+    if (!ConnectionService.INSTANCE) {
+      ConnectionService.INSTANCE = new ConnectionService();
+    }
+
+    return ConnectionService.INSTANCE;
+  };
+
+  lock = async (): Promise<boolean> => this.onLock(this.onClean);
+
+  private onClean = (): boolean => {
+    this.connections.clear();
+    return true;
+  };
+
+  unlock = async (): Promise<boolean> => {
+    const encryped = await this.connectionsStorage.get<string>();
+
+    if (encryped) {
+      const decrypted = this.cryptoService.decrypt(encryped, { mode: ECryptMode.MNEMONIC });
+      this.connections = new Map(JSON.parse(decrypted) as Iterable<[string, IIdenityConnection]>);
+    }
+
+    this.isUnlocked = true;
+    this.onUnlocked();
+
+    return true;
+  };
+
+  connect = async ({ commitment }: IConnectArgs, { urlOrigin }: IZkMetadata): Promise<void> => {
+    if (!urlOrigin) {
+      throw new Error("CryptKeeper: origin is not provided");
+    }
+
+    const identity = this.zkIdenityService.getIdentity(commitment);
+
+    if (!identity) {
+      throw new Error("CryptKeeper: identity is not found");
+    }
+
+    const connection = this.getIdentityConnection(commitment, identity.metadata);
+    this.connections.set(urlOrigin, connection);
+    await this.save();
+
+    await this.browserController.pushEvent(
+      { type: EventName.CONNECT_IDENTITY, payload: omit(connection, ["commitment"]) },
+      { urlOrigin },
+    );
+  };
+
+  disconnect = async (payload: unknown, { urlOrigin }: IZkMetadata): Promise<void> => {
+    await this.isOriginConnected(payload, { urlOrigin });
+    this.connections.delete(urlOrigin!);
+    await this.save();
+
+    await this.browserController.pushEvent(
+      { type: EventName.DISCONNECT_IDENTITY, payload: { urlOrigin } },
+      { urlOrigin },
+    );
+  };
+
+  isOriginConnected = (payload: unknown, { urlOrigin }: IZkMetadata): unknown => {
```

Comment:
Why do we need `payload` param for this function?

Reply:
Because it's a middleware and payload must be passed though it.

Reply:
Until now it is not a middleware its only a check for `urlOrigin`  inside `disconnect`
`const isConnected = this.connections.has(urlOrigin);`

And it throws only errors in cases of not connected or no `urlOrigin` it desn't deal with Payload and doesn't block Payload. 

Or do you intent to use it outside in CryptKeeper controller handler as a middleware?

Reply:
In my current branch I use it as a middleware for handler. For this case I need to pass payload though this chain of checks without changing it.

```
this.handler.add(
      RPCExternalAction.GENERATE_SEMAPHORE_PROOF,
      this.lockService.ensure,
      this.approvalService.isOriginApproved,
      this.connectionService.isOriginConnected,
      this.protocolService.generateSemaphoreProof,
    );
```